RuntimeException in MediaService if Play Services are missing

**Describe the bug**

On devices without Play Services, the `MediaService` crashes on every network change.

**To Reproduce**
<!-- Steps to reproduce the behavior: -->
1. Install Jellyfin Android on a device without Play Services
2. Log in, watch some movies
3. Close Jellyfin, reboot, get on with life
4. Change networks, e.g. from 4G to WiFi
5. Now the  `MediaService` seems to be fired, crashes, and Android greets you with the "Jellyfin keeps stopping" dialog.
 
The annoying thing here is, that you get A LOT of those "Jellyfin keeps stopping" popups as you move throughout the day (not actively using Jellyfin at all).

I don't have a device to cast to, so I don't know if casting works at all (maybe it crashes, too). That might be worth a separate issue, right now I'm most concerned about Jellyfin crashes interrupting my unrelated phone usage.

**Expected behavior**

Jellyfin should gracefully handle the absence of Play Services.

**Logs**
```
FATAL EXCEPTION: main
Process: org.jellyfin.mobile, PID: 18244

java.lang.RuntimeException: Unable to create service org.jellyfin.mobile.media.MediaService: java.lang.RuntimeException: com.google.android.gms.cast.framework.zzad: com.google.android.gms.dynamite.DynamiteModule$LoadingException: No acceptable module found. Local version is 0 and remote version is 0.
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3965)
	at android.app.ActivityThread.access$1500(ActivityThread.java:219)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1875)
	at android.os.Handler.dispatchMessage(Handler.java:107)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7356)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)

Caused by: java.lang.RuntimeException: com.google.android.gms.cast.framework.zzad: com.google.android.gms.dynamite.DynamiteModule$LoadingException: No acceptable module found. Local version is 0 and remote version is 0.
	at j.b.a.c.c.s.b.c(com.google.android.gms:play-services-cast-framework@@19.0.0:10)
	at org.jellyfin.mobile.cast.CastPlayerProvider.<init>(CastPlayerProvider.kt:2)
	at org.jellyfin.mobile.media.MediaService$castPlayerProvider$2.invoke(MediaService.kt:1)
	at n.h.getValue(LazyJVM.kt:5)
	at org.jellyfin.mobile.media.MediaService.getCastPlayerProvider(Unknown Source:2)
	at org.jellyfin.mobile.media.MediaService.onCreate(MediaService.kt:66)
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3953)
	... 8 more

Caused by: com.google.android.gms.cast.framework.zzad: com.google.android.gms.dynamite.DynamiteModule$LoadingException: No acceptable module found. Local version is 0 and remote version is 0.
	at j.b.a.c.h.b.h.a(com.google.android.gms:play-services-cast-framework@@19.0.0:10)
	at j.b.a.c.c.s.b.<init>(com.google.android.gms:play-services-cast-framework@@19.0.0:8)
	at j.b.a.c.c.s.b.c(com.google.android.gms:play-services-cast-framework@@19.0.0:9)
	... 14 more

Caused by: com.google.android.gms.dynamite.DynamiteModule$LoadingException: No acceptable module found. Local version is 0 and remote version is 0.
	at com.google.android.gms.dynamite.DynamiteModule.a(com.google.android.gms:play-services-basement@@17.1.0:27)
	at j.b.a.c.h.b.h.a(com.google.android.gms:play-services-cast-framework@@19.0.0:2)
	... 16 more
```

**Screenshots**

<img src=https://user-images.githubusercontent.com/33295590/108033706-b3146280-7034-11eb-99e1-3bf1e3bb505a.png width=400>

**System:**
 - App version: 2.2.3
 - Android version: Android 10, LineageOS 17.1 (no GApps - you guess it)
 - Device manufacturer and model: OnePlus 5 (cheeseburger)
 - Jellyfin server version: 10.6.4

 - The type of video player implementation used: integrated player (ExoPlayer) => thank you, the more native player is awesome!

**Additional context**
N/A

--- a/app/src/libre/java/org/jellyfin/mobile/cast/CastPlayerProvider.kt
+++ b/app/src/libre/java/org/jellyfin/mobile/cast/CastPlayerProvider.kt
@@ -6,5 +6,5 @@ import org.jellyfin.mobile.media.MediaService
 class CastPlayerProvider(@Suppress("UNUSED_PARAMETER") mediaService: MediaService) : ICastPlayerProvider {
     override val isCastSessionAvailable: Boolean = false
 
-    override fun get(): Player = throw NotImplementedError("CastPlayer isn't implemented in libre flavor")
+    override fun get(): Player? = null
 }

--- a/app/src/main/java/org/jellyfin/mobile/cast/ICastPlayerProvider.kt
+++ b/app/src/main/java/org/jellyfin/mobile/cast/ICastPlayerProvider.kt
@@ -5,5 +5,5 @@ import com.google.android.exoplayer2.Player
 interface ICastPlayerProvider {
     val isCastSessionAvailable: Boolean
 
-    fun get(): Player
+    fun get(): Player?
 }

--- a/app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
+++ b/app/src/main/java/org/jellyfin/mobile/media/MediaService.kt
@@ -135,7 +135,7 @@ class MediaService : MediaBrowserServiceCompat() {
 
         switchToPlayer(
             previousPlayer = null,
-            newPlayer = if (castPlayerProvider.isCastSessionAvailable) castPlayerProvider.get() else exoPlayer
+            newPlayer = if (castPlayerProvider.isCastSessionAvailable) castPlayerProvider.get()!! else exoPlayer
         )
         notificationManager.showNotificationForPlayer(currentPlayer)
 
@@ -162,7 +162,7 @@ class MediaService : MediaBrowserServiceCompat() {
         clientPackageName: String,
         clientUid: Int,
         rootHints: Bundle?
-    ): BrowserRoot? = libraryBrowser.getRoot(rootHints)
+    ): BrowserRoot = libraryBrowser.getRoot(rootHints)
 
     override fun onLoadChildren(parentId: String, result: Result<List<MediaItem>>) {
         result.detach()
@@ -199,12 +199,15 @@ class MediaService : MediaBrowserServiceCompat() {
             exoPlayer.setMediaItems(mediaItems)
             exoPlayer.prepare()
             exoPlayer.seekTo(initialPlaybackIndex, playbackStartPositionMs)
-        } else /* currentPlayer == castPlayer */ {
-            castPlayerProvider.get().setMediaItems(
-                mediaItems,
-                initialPlaybackIndex,
-                playbackStartPositionMs,
-            )
+        } else {
+            val castPlayer = castPlayerProvider.get()
+            if (currentPlayer == castPlayer) {
+                castPlayer.setMediaItems(
+                    mediaItems,
+                    initialPlaybackIndex,
+                    playbackStartPositionMs,
+                )
+            }
         }
     }
 
@@ -241,7 +244,8 @@ class MediaService : MediaBrowserServiceCompat() {
 
     @Suppress("unused")
     fun onCastSessionAvailable() {
-        switchToPlayer(currentPlayer, castPlayerProvider.get())
+        val castPlayer = castPlayerProvider.get() ?: return
+        switchToPlayer(currentPlayer, castPlayer)
     }
 
     @Suppress("unused")

--- a/app/src/proprietary/java/org/jellyfin/mobile/cast/CastPlayerProvider.kt
+++ b/app/src/proprietary/java/org/jellyfin/mobile/cast/CastPlayerProvider.kt
@@ -6,24 +6,26 @@ import com.google.android.exoplayer2.ext.cast.SessionAvailabilityListener
 import com.google.android.gms.cast.framework.CastContext
 import org.jellyfin.mobile.media.MediaService
 
-class CastPlayerProvider(mediaService: MediaService) : ICastPlayerProvider {
-    private val castPlayer: CastPlayer = CastPlayer(CastContext.getSharedInstance(mediaService)).apply {
-        setSessionAvailabilityListener(CastSessionAvailabilityListener(mediaService))
-        addListener(mediaService.playerListener)
+class CastPlayerProvider(private val mediaService: MediaService) : ICastPlayerProvider, SessionAvailabilityListener {
+    private val castPlayer: CastPlayer? = try {
+        CastPlayer(CastContext.getSharedInstance(mediaService)).apply {
+            setSessionAvailabilityListener(this@CastPlayerProvider)
+            addListener(mediaService.playerListener)
+        }
+    } catch (e: Exception) {
+        null
     }
 
     override val isCastSessionAvailable: Boolean
-        get() = castPlayer.isCastSessionAvailable
+        get() = castPlayer?.isCastSessionAvailable == true
 
-    override fun get(): Player = castPlayer
+    override fun get(): Player? = castPlayer
 
-    private class CastSessionAvailabilityListener(private val mediaService: MediaService) : SessionAvailabilityListener {
-        override fun onCastSessionAvailable() {
-            mediaService.onCastSessionAvailable()
-        }
+    override fun onCastSessionAvailable() {
+        mediaService.onCastSessionAvailable()
+    }
 
-        override fun onCastSessionUnavailable() {
-            mediaService.onCastSessionUnavailable()
-        }
+    override fun onCastSessionUnavailable() {
+        mediaService.onCastSessionUnavailable()
     }
 }

--- a/app/src/proprietary/java/org/jellyfin/mobile/cast/Chromecast.java
+++ b/app/src/proprietary/java/org/jellyfin/mobile/cast/Chromecast.java
@@ -175,6 +175,7 @@ public final class Chromecast implements IChromecast {
     public boolean setup(JavascriptCallback javascriptCallback) {
         this.eventCallback = javascriptCallback;
         // Ensure any existing scan is stopped
+        if (connection == null) return false;
         connection.stopRouteScan(clientScan, () -> {
             if (scanCallback != null) {
                 scanCallback.error(ChromecastUtilities.createError("cancel", "Scan stopped because setup triggered."));
@@ -196,6 +197,7 @@ public final class Chromecast implements IChromecast {
      * @return true for cordova
      */
     public boolean initialize(final String appId, String autoJoinPolicy, String defaultActionPolicy, final JavascriptCallback javascriptCallback) {
+        if (connection == null) return false;
         connection.initialize(appId, javascriptCallback);
         return true;
     }
@@ -210,6 +212,7 @@ public final class Chromecast implements IChromecast {
      * @return true for cordova
      */
     public boolean requestSession(final JavascriptCallback javascriptCallback) {
+        if (connection == null) return false;
         connection.requestSession(new ChromecastConnection.RequestSessionCallback() {
             @Override
             public void onJoin(JSONObject jsonSession) {
@@ -238,6 +241,7 @@ public final class Chromecast implements IChromecast {
      * @return true for cordova
      */
     public boolean selectRoute(final String routeId, final JavascriptCallback javascriptCallback) {
+        if (connection == null) return false;
         connection.selectRoute(routeId, new ChromecastConnection.SelectRouteCallback() {
             @Override
             public void onJoin(JSONObject jsonSession) {
@@ -479,6 +483,7 @@ public final class Chromecast implements IChromecast {
      * @return true for cordova
      */
     public boolean sessionStop(JavascriptCallback javascriptCallback) {
+        if (connection == null) return false;
         connection.endSession(true, javascriptCallback);
         return true;
     }
@@ -490,6 +495,7 @@ public final class Chromecast implements IChromecast {
      * @return true for cordova
      */
     public boolean sessionLeave(JavascriptCallback javascriptCallback) {
+        if (connection == null) return false;
         connection.endSession(false, javascriptCallback);
         return true;
     }
@@ -503,6 +509,7 @@ public final class Chromecast implements IChromecast {
      * @return true for cordova
      */
     public boolean startRouteScan(JavascriptCallback javascriptCallback) {
+        if (connection == null) return false;
         if (scanCallback != null) {
             scanCallback.error(ChromecastUtilities.createError("cancel", "Started a new route scan before stopping previous one."));
         }
@@ -537,6 +544,7 @@ public final class Chromecast implements IChromecast {
      * @return true for cordova
      */
     public boolean stopRouteScan(JavascriptCallback javascriptCallback) {
+        if (connection == null) return false;
         // Stop any other existing clientScan
         connection.stopRouteScan(clientScan, () -> {
             if (scanCallback != null) {
@@ -573,7 +581,9 @@ public final class Chromecast implements IChromecast {
         sessionStop(callback);
         media.destroy();
         media = null;
-        connection.destroy();
+        if (connection != null) {
+            connection.destroy();
+        }
         connection = null;
         eventCallback = null;
     }

